File: nim/protein-translation/protein_translation.nim
func chunked(s: string, chunkLen: int): seq[string] =
  for i in countup(0, s.high, chunkLen):
    result.add s[i .. min(s.high, i + chunkLen - 1)]

func toProtein(codon: string): string =
  case codon:
    of "AUG": "Methionine"
    of "UUC", "UUU": "Phenylalanine"
    of "UUA", "UUG": "Leucine"
    of "UCU", "UCC", "UCA", "UCG": "Serine"
    of "UAU", "UAC": "Tyrosine"
    of "UGU", "UGC": "Cysteine"
    of "UGG": "Tryptophan"
    of "UAA", "UAG", "UGA": "STOP"
    else: raise newException(KeyError, "invalid codon")

proc translate*(rna: string): seq[string] =
  for codon in rna.chunked 3:
    let protein = codon.toProtein
    if protein == "STOP":
      break
    result.add protein

File: r/protein-translation/protein-translation.R
dict = list(
    AUG = "Methionine", UUU = "Phenylalanine", UUC = "Phenylalanine",
    UUA = "Leucine", UUG = "Leucine", UCU = "Serine", UCC = "Serine",
    UCA = "Serine", UCG = "Serine", UAU = "Tyrosine", UAC = "Tyrosine",
    UGU = "Cysteine", UGC = "Cysteine", UGG = "Tryptophan", UAA = "STOP",
    UAG = "STOP", UGA = "STOP")

translate = function(bases) {
  if (nchar(bases) == 0) return(NULL)
  codons = strsplit(bases, "(?<=.{3})", perl = TRUE)[[1]]
  output = c()
  for (codon in codons) {
    if (nchar(codon) != 3) stop("Invalid codon")
    amino_acid = dict[[codon]]
    if (is.null(amino_acid)) stop("Invalid codon")
    if (amino_acid == "STOP") break
    output = append(output, amino_acid)
  }
  output
}

